### PR TITLE
Fix key list lookup with hash collisions

### PR DESCRIPTION
**Pull request by dimas-b**
This change fixes key lookup when hash collisions cause the key to be stored
in the same segments, in front of its natural position.

### CODE REVIEW

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java [hunk 1]
```diff
@@ -524,4 +524,39 @@ private static void checkKeysAndValuesIndividually(
           .isEqualTo(valueGen.apply(i).serialized());
     }
   }
+
+  public static Stream<List<String>> progressiveKeyNames() {
+    return IntStream.range(1, 100)
+        .mapToObj(i -> IntStream.range(1, i).mapToObj(j -> "a-" + j).collect(Collectors.toList()));
```
Comment (snazy) [inline versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java:530]: You probably want `rangeClosed` here

**Comment by snazy** [comment]
Ouch. Could be related to String.hash collisions and overflow handling maybe. Probably worth to port this test to `TestFetchValuesUsingOpenAddressing`/`TestKeyListBuildState`, so it doesn't depend on the DB-adapter default configurations.
Did a quick experiment, and a load-factor of 0.45 instead of 0.65 doesn't trigger the issue with this test (not saying that 0.45 fixes this). As an idea: maybe try with higher load factors.

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java (hunk 1 above)
Comment (dimas-b) [reply versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java]: 1 to 99 is enough, 100 is not a hard limit more of a "large enough" number :shrug:

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 2]
```diff
@@ -84,54 +85,17 @@ public void segmentWrapAroundWithAbsentKey() {
...
     fetch.checkForKeys(2, singleton, keyHits);
-    fetch.entityLoaded(getKeyListEntity(2, 2));
+    fetch.entityLoaded(getKeyListEntity(4, 2, 2, KeyWithSpecificHash::withMaxIntHash));
     fetch.checkForKeys(3, singleton, keyHits);
     verifyNoInteractions(keyHits);
   }
 
-  /** Test finding extant keys pushed out of the final (highest-indexed) segment by collisions. */
-  @Test
-  public void segmentWrapAroundWithPresentKey() {
```
Comment (dimas-b) [inline versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java:100]: I hope the `keyCollisions` test covers this now

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java [hunk 3]
```diff
@@ -524,4 +524,56 @@ private static void checkKeysAndValuesIndividually(
           .isEqualTo(valueGen.apply(i).serialized());
     }
   }
+
+  public static Stream<List<String>> progressivelyManyKeyNames() {
+    // The seed of `7` was chosen after a few manual tries because it caused the most test failures
+    // in the key lookup code before the commit where this test was introduced.
```
Comment (snazy) [inline versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java:530]: Can you add the PR number instead of "the commit" ?

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java [hunk 4]
```diff
@@ -524,4 +524,56 @@ private static void checkKeysAndValuesIndividually(
...
+                IntStream.range(1, i)
+                    .mapToObj(j -> Long.toString(Math.abs(random.nextLong()), Character.MAX_RADIX))
+                    .collect(Collectors.toList()));
+  }
+
+  /**
+   * Validates key construction and lookup with random, yet deterministic key names hoping to cover
+   * all edge cases and hash collisions by using a somewhat large test data set.
+   *
+   * <p>Note: the number of keys that is effective for validating the hash lookup indirectly depends
```
Comment (snazy) [inline versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java:546]: Hint: You can get a `DatabaseAdapter` instance here as a method parameter, using the `@NessieDbAdapter` plus two `@NessieDbAdapterConfigItem` annotations for the method parameter to ensure the test's using the correct adapter configuration.

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 5]
```diff
@@ -199,6 +165,105 @@ public void segmentPrefetching() {
...
+                .isEqualTo(movedKeyFoundInRound);
+          } else { // Ordinary keys are found on the first attempt
+            assertThat(r).describedAs("Expected round for key " + key).isEqualTo(0);
+          }
+          break;
+        }
+      }
+
+      // Make sure each key is found
+      assertThat(hit.get()).describedAs("Key " + key).isNotNull();
```
Comment (snazy) [inline versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java:262]: Nit: can be continued w/ `.extracting(KeyListEntry::getKey).isEqualTo(key);` (+ removing the next line).
Maybe better: Use the `@ExtendWith(SoftAssertionsExtension.class)` + an instance field `@InjectSoftAssertions SoftAssertions softly` and use `softly.assertThat`. Helps a lot when using many assertThat in a single test

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 6]
```diff
@@ -140,8 +140,16 @@ Collection<Key> checkForKeys(
...
+    // then return an empty collection early. Note that the initial segment may need to be examined
+    // twice when the key is moved in front of its natural position due to key collisions. This
+    // is achieved by subtracting 1 from `round`. On the first round the initial segment is scanned
+    // from the "natural" key position. On the last round the initial segment is scanned from index
+    // zero. This may incur some extra key comparisons if the segment is full and the scanning
+    // overtakes the "natural" key position on the last round. However, this is probably a rare case
+    // and not worth complicating the lookup logic to avoid it. In any case, the double lookup in
+    // the initial segment does not cause extra I/O because the segment is pre-fetched by the time
+    // it is reused.
+    if (keyListsArray.length <= round - 1) {
```
Comment (snazy) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:152]: Maybe just `< round` to avoid the subtraction ?

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 7]
```diff
@@ -140,8 +140,16 @@ Collection<Key> checkForKeys(
       int round, Collection<Key> remainingKeys, Consumer<KeyListEntry> resultConsumer) {
     List<Key> keysForNextRound = new ArrayList<>();
     // If we've already examined every segment (implying a miss on a completely full hashmap),
-    // then return an empty collection early
-    if (keyListsArray.length <= round) {
+    // then return an empty collection early. Note that the initial segment may need to be examined
```
Comment (snazy) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:143]: THanks for the explanation here! This really helps to understand what's happening.

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 8]
```diff
@@ -279,20 +347,15 @@ private static String intToPaddedHexString(int i) {
...
+  private static KeyListEntity getKeyListEntity(
+      int startingBucket, int segmentIndex, int entryCount, Function<Integer, Key> bucketToKey) {
     ImmutableKeyList.Builder listBuilder = ImmutableKeyList.builder();
 
+    int bucket = startingBucket;
     for (int i = 0; i < entryCount; i++) {
-      String keyString = prefix + i;
-      listBuilder.addKeys(getKeyListEntry(keyString));
+      listBuilder.addKeys(getKeyListEntry(bucketToKey.apply(bucket)));
+      bucket++;
```
Comment (snazy) [inline versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java:358]: Super-nit: move into the for-loop

**Review by snazy — COMMENTED** [review]
LGTM - few minor comment.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java (hunk 6 above)
Comment (dimas-b) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:152]: Sure.

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java (hunk 3 above)
Comment (dimas-b) [reply versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java:530]: done

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java (hunk 5 above)
Comment (dimas-b) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java:262]: Converted to `.extracting`

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java (hunk 8 above)
Comment (dimas-b) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java:358]: done (I do not even recall when I used multiple variables in a `for` loop in java last time :sweat_smile: )

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java (hunk 4 above)
Comment (dimas-b) [reply versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java:546]: updated

**Review by snazy — DISMISSED** [review]
LGTM, maybe let @dalaro cross check the tests.

**Comment by dimas-b** [comment]
rebased

**Review by snazy — APPROVED** [review]

**Comment by dalaro** [comment]
I think #5174 currently contains a superset of the commits in this PR.  So, although I read the comments and conversation here for background, I've reviewed there.  However, if you'd rather I post my review here, that's no problem, just let me know.

**Comment by dimas-b** [comment]
I hit these two failure modes separately and opened two PRs initially. I feel like abandoning this PR in favour of #5174 now. I'll squash the two commit under #5174.
